File: Server/src/main/content/global/interaction/item/plugins/TeleportCrystalPlugin.java
package content.global.interaction.item.plugins;

import core.cache.def.impl.ItemDefinition;
import core.game.dialogue.Dialogue;
import core.game.interaction.OptionHandler;
import core.game.node.Node;
import core.game.node.entity.npc.NPC;
import core.game.node.entity.player.Player;
import core.game.node.entity.player.link.TeleportManager.TeleportType;
import core.game.node.item.Item;
import core.game.world.map.Location;
import core.game.world.map.zone.impl.WildernessZone;
import core.plugin.Initializable;
import core.plugin.Plugin;

import static core.api.ContentAPIKt.hasRequirement;

/**
 * The Teleport crystal plugin.
 */
@Initializable
public final class TeleportCrystalPlugin extends OptionHandler {

    @Override
    public Plugin<Object> newInstance(Object arg) throws Throwable {
        ItemDefinition.forId(6099).getHandlers().put("option:activate", this);
        ItemDefinition.forId(6100).getHandlers().put("option:activate", this);
        ItemDefinition.forId(6101).getHandlers().put("option:activate", this);
        ItemDefinition.forId(6102).getHandlers().put("option:activate", this);
        new TeleportCrystalDialogue().init();
        return this;
    }

    @Override
    public boolean handle(Player player, Node node, String option) {
        if (!hasRequirement(player, "Mourning's End Part I")) return true;
        if (!WildernessZone.checkTeleport(player, 20)) {
            player.getPacketDispatch().sendMessage("The crystal is unresponsive.");
            return true;
        }
        player.getDialogueInterpreter().open(TeleportCrystalDialogue.ID, 1, node.asItem().getId());
        //degrade(player, node.asItem());
        return true;
    }


    @Override
    public boolean isWalk() {
        return false;
    }


    /**
     * The Teleport crystal dialogue.
     */
    public static final class TeleportCrystalDialogue extends Dialogue {


        /**
         * The constant ID.
         */
        public static final int ID = 3999111;
        private Integer itemId;


        /**
         * Instantiates a new Teleport crystal dialogue.
         */
        public TeleportCrystalDialogue() {

        }


        /**
         * Instantiates a new Teleport crystal dialogue.
         *
         * @param player the player
         */
        public TeleportCrystalDialogue(Player player) {
            super(player);
        }

        @Override
        public Dialogue newInstance(Player player) {
            return new TeleportCrystalDialogue(player);
        }

        @Override
        public boolean open(Object... args) {
            itemId = (Integer) args[1];
            switch ((Integer) args[0]) {
                case 1:
                    interpreter.sendOptions("Select an Option", "Teleport to Lletya", "Cancel");
                    stage = 100;
                    break;
            }
            return true;
        }

        @Override
        public boolean handle(int interfaceId, int buttonId) {
            NPC npc = new NPC(2376);
            NPC billTeach = new NPC(3155);
            switch (stage) {
                case 100:
                    switch (buttonId) {
                        case 1:
                            player.getTeleporter().send(new Location(2329, 3172), TeleportType.NORMAL);
                            degrade(player, new Item(itemId));
                            break;
                        case 2:
                            end();
                            break;
                    }
                    break;
            }
            return true;
        }


        private static void degrade(Player p, Item item) {
            int id = item.getId();
            int newItem = item.getId() + 1;
            if (id < 6102) {
                p.getInventory().remove(new Item(id, 1));
                p.getInventory().add(new Item(newItem, 1));
                p.getPacketDispatch().sendMessage("Your teleportation crystal has degraded from use.");
            } else {
                p.getInventory().remove(new Item(id, 1));
                p.getInventory().add(new Item(newItem, 1));
                p.getPacketDispatch().sendMessages("Your teleportation crystal has degraded to a tiny elf crystal,", "Eluned can re-enchant it.");
            }
        }

        @Override
        public int[] getIds() {
            return new int[]{ID};
        }

    }

}

File: Server/src/main/content/global/handlers/item/plugin/TeleportCrystalPlugin.kt
package content.global.handlers.item.plugin

import core.api.addItem
import cfg.consts.Items
import core.api.hasRequirement
import core.api.removeItem
import core.api.sendMessage
import core.cache.def.impl.ItemDefinition
import core.game.dialogue.Dialogue
import core.game.interaction.OptionHandler
import core.game.node.Node
import core.game.node.entity.npc.NPC
import core.game.node.entity.player.Player
import core.game.node.entity.player.link.TeleportManager.TeleportType
import core.game.node.item.Item
import core.game.world.map.Location
import core.game.world.map.zone.impl.WildernessZone
import core.plugin.Initializable
import core.plugin.Plugin

/**
 * Teleport crystal plugin.
 */
@Initializable
class TeleportCrystalPlugin : OptionHandler() {

    override fun newInstance(arg: Any?): Plugin<Any> {
        ItemDefinition.forId(Items.TELEPORT_CRYSTAL_4_6099).handlers["option:activate"] = this
        ItemDefinition.forId(Items.TELEPORT_CRYSTAL_3_6100).handlers["option:activate"] = this
        ItemDefinition.forId(Items.TELEPORT_CRYSTAL_2_6101).handlers["option:activate"] = this
        ItemDefinition.forId(Items.TELEPORT_CRYSTAL_1_6102).handlers["option:activate"] = this
        TeleportCrystalDialogue().init()
        return this
    }

    override fun handle(player: Player, node: Node, option: String): Boolean {
        if (!hasRequirement(player, "Mourning's End Part I")) return true
        if (!WildernessZone.checkTeleport(player, 20)) {
            sendMessage(player, "The crystal is unresponsive.")
            return true
        }
        player.dialogueInterpreter.open(TeleportCrystalDialogue.ID, 1, node.asItem().id)
        //degrade(player, node.asItem());
        return true
    }


    override fun isWalk(): Boolean {
        return false
    }

    /**
     * Teleport crystal dialogue.
     */
    class TeleportCrystalDialogue(player: Player? = null) : Dialogue(player) {
        private var itemId: Int? = null

        override fun newInstance(player: Player): Dialogue {
            return TeleportCrystalDialogue(player)
        }

        override fun open(vararg args: Any): Boolean {
            itemId = args[1] as Int
            when (args[0] as Int) {
                1 -> options("Teleport to Lletya", "Cancel").also { stage = 100 }
            }
            return true
        }

        override fun handle(interfaceId: Int, buttonId: Int): Boolean {
            val npc = NPC(2376)
            val billTeach = NPC(3155)
            when (stage) {
                100 -> when (buttonId) {
                    1 -> {
                        player.teleporter.send(Location(2329, 3172), TeleportType.NORMAL)
                        degrade(player, Item(itemId!!))
                    }

                    2 -> end()
                }
            }
            return true
        }


        override fun getIds(): IntArray {
            return intArrayOf(ID)
        }

        companion object {
            const val ID: Int = 3999111
            private fun degrade(player: Player, item: Item) {
                val id = item.id
                val newItem = item.id + 1
                if (id < 6102) {
                    removeItem(player, Item(id, 1))
                    addItem(player, newItem, 1)
                    sendMessage(player, "Your teleportation crystal has degraded from use.")
                } else {
                    removeItem(player, Item(id, 1))
                    addItem(player, newItem, 1)
                    sendMessage(player, "Your teleportation crystal has degraded to a tiny elf crystal,")
                    sendMessage(player, "Eluned can re-enchant it.")
                }
            }
        }
    }
}
